on_device_training/mobilevit/mobilevit_console/mobilevit_console/Program.cs

```diff
@@ -0,0 +1,164 @@
+        /// no trimming will happen and the full dataset will be used for training.</param>
+        /// <returns></returns>
+        public static DataLoader loadDataset(string pathToFER, int trimDatasetTo = -1)
+        {
+            var watch = System.Diagnostics.Stopwatch.StartNew();
+            DataLoader dl = new DataLoader();
+            dl.loadFER(pathToFER);
+            if (trimDatasetTo > 0)
+            {
+                dl.trimFER(trimDatasetTo);
+            }
+
+            watch.Stop();
+            Console.WriteLine($"Execution Time for loading images and shuffling: {watch.ElapsedMilliseconds} ms");
+
+            return dl;
+        }
+
+        public static void CreateAndRunTrainingSession(int numEpochs, int batchSize, DataLoader dl)
+        {
+            string parentDir = Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.Parent.Parent.FullName;
+
+            var state = CheckpointState.LoadCheckpoint(CHECKPOINTPATH);
+
+            TrainingSession ts = new TrainingSession(state, TRAININGMODELPATH, EVALMODELPATH, OPTIMIZERMODELPATH);
+            train(ts, dl, numEpochs, batchSize);
+
+            Console.WriteLine("############################################################### training finished");
+
+            var outputNames = new List<string> { "outputs" };
+
+            ts.ExportModelForInferencing(TRAINEDMODELFILE, outputNames);
+        }
+
+        public static void runInferenceSession(InferenceSession inferenceSession, DataLoader dl, string pathToInference)
+        {
+            Console.WriteLine($"Inferencing now for {pathToInference}");
+
+            var input = dl.imageProcessingForInference(pathToInference);
+            var inputs = new List<NamedOnnxValue> { NamedOnnxValue.CreateFromTensor("pixel_values", input) };
+
+            foreach (var r in inferenceSession.Run(inputs))
+            {
+                var resultsTensor = r.AsTensor<float>();
+
+                printPredictedEmotion(resultsTensor);
+
+                Console.WriteLine(string.Join(",     ", DataLoader.EMOTIONSLABELS));
+                Console.WriteLine(resultsTensor.GetArrayString());
+            }
+        }
+
+        public static float train(TrainingSession ts, DataLoader dl, int numEpochs, int batchSize)
+        {
+            float loss = 0;
+            for (int i = 0; i < numEpochs; i++)
+            {
+                loss = trainEpoch(ts, dl, batchSize);
+
+                Console.WriteLine($"Loss after epoch {i}: {loss}");
+            }
+
+            return loss;
+        }
+
+        public static float trainEpoch(TrainingSession ts, DataLoader dl, int batchSize)
+        {
+            int steps = dl.getNumSteps(batchSize);
+            // int steps = 20;
+            Console.WriteLine("steps: " + steps);
+
+            float loss = 0;
+
+            for (int step = 0; step < steps; step++)
+            {
+                var watch = System.Diagnostics.Stopwatch.StartNew();
```

Comment:
move this below the preprocessing step? This way we will know the time for training steps